contracts/access/AccessControlAdminRules.sol

```diff
@@ -0,0 +1,220 @@
+ *
+ * This contract implements the following risk mitigations on top of the AccessControl implementation:
+ *
+ * * Only one account holds the `DEFAULT_ADMIN_ROLE` at every time after construction except when renounced.
+ * * Enforce a 2-step process to transfer the `DEFAULT_ADMIN_ROLE` to another account.
+ *   - Even when it's been renounced.
+ * * Enforce a configurable delay between the two steps, with the ability to cancel in between.
+ *   - Even after the timer has passed to avoid locking it forever.
+ * * The `DEFAULT_ADMIN_ROLE` 's admin can be only held by itself.
+ *
+ * Once you understand what the {constructor} parameters, you can use this reference implementation:
+ *
+ * ```solidity
+ * contract MyToken is AccessControlAdminRules {
+ *   constructor() AccessControlAdminRules(
+ *     60 * 60 * 24, // 3 days
+ *     _msgSender() // Explicit initial `DEFAULT_ADMIN_ROLE` holder
+ *    ) {}
+ *}
+ * ```
+ *
+ * NOTE: The `delay` is only configurable in the constructor to avoid issues related with handling
+ * delay management during the transfer is pending to be completed.
+ *
+ * _Available since v4.9._
+ */
+abstract contract AccessControlAdminRules is IAccessControlAdminRules, AccessControl {
+    uint48 private immutable _delay;
+
+    address private _currentAdmin;
+    uint48 private _delayedUntil;
+
+    address private _pendingAdmin;
+
+    /**
+     * @dev Sets the initial values the internal delay and {owner}.
+     *
+     * The `delay` value is immutable. It can only be set at the constructor.
+     */
+    constructor(uint48 delay, address initialAdmin) {
+        _delay = delay;
+        _grantRole(DEFAULT_ADMIN_ROLE, initialAdmin);
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-owner}
+     */
+    function owner() public view virtual returns (address) {
+        return _currentAdmin;
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-delayedUntil}
+     */
+    function delayedUntil() public view virtual returns (uint48) {
+        return _delayedUntil;
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-pendingAdmin}
+     */
+    function pendingAdmin() public view virtual returns (address) {
+        return _pendingAdmin;
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(IAccessControlAdminRules).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-beginAdminTransfer}
+     */
+    function beginAdminTransfer(address newAdmin) public override onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(delayedUntil() == 0, "AccessControl: pending admin already set");
+        _delayedUntil = SafeCast.toUint48(block.timestamp) + _delay;
+        _pendingAdmin = newAdmin;
+        emit AdminRoleChangeStarted(pendingAdmin(), delayedUntil());
```

Comment:
That would be cheaper (avoid 2 sloads). It raises the question of "how to deal with overrides".
```suggestion
        emit AdminRoleChangeStarted(_pendingAdmin, _delayedUntil);
```

Reply:
Note, right now the functions are not virtual, so the overriding is not a valid reason to do the call

Reply:
> I removed them, but, should `IAccessControlAdminRules` functions be virtual? thinking

The naming suggest its an interface, so all functions should be virtual.

Reply:
It is an interface. Sorry, I'll rephrase:

> should `IAccessControlAdminRules` functions implemented in `AccessControlAdminRules` be virtual?

This is the discussion I started internally. I made them virtual.